network/network_interface.ml

```diff
@@ -25,299 +25,609 @@ let uri () = "file:" ^ !default_path
+          "the infomation about the script and the error"]]
+  | Read_error of string
+        [@doc [
+          "[Read_error (file)] is reported ";
+          "when error occurs when reading [file]"]]
+  | Write_error of string
+        [@doc [
+          "[Write_error (file)] is reported ";
+          "when error occurs when writing [file]"]]
+  | Not_implemented
+      [@doc [
+        "[Not_implemented] is reported ";
+        "if the interface is not implemented"]]
+  | Vlan_in_use of (string * int)
+        [@doc [
+          "[Vlan_in_use (bridge, vlan_id)] is reported ";
+          "when [vlan_id] on [bridge] is inuse"]]
+  | PVS_proxy_connection_error
+      [@doc [
+        "[PVS_proxy_connection_error] is reported ";
+        "when unable to connect PVS proxy"]]
+  | Unknown_error
+      [@doc [
+        "The default variant for forward compatibility."
+      ]]
+[@@default Unknown_error]
+[@@deriving rpcty]
+
+exception NetworkError of errors
+
+let err = Error.{
+    def = errors;
+    raiser = (function | e -> raise (NetworkError e));
+    matcher = function | NetworkError e -> Some e | _ -> None
+  }
+
+module API(R : RPC) = struct
+  open R
+
+  let description = Interface.{
+      name = "Network";
+      namespace = None;
+      description = [
+        "This interface is used by Xapi and networkd to manage ";
+        "Xenserver network bridges and devices .";
+      ];
+      version=(1,0,0);
+    }
+
+  let implementation = implement description
+
+  let debug_info_p = Param.mk ~description:[
+      "an uninterpreted string to associate with the operation."
+    ] Types.string
+
+  let unit_p = Param.mk Types.unit
+
+  (* external clear_state: unit -> unit = "" *)
+  let clear_state = declare
+      "clear_state"
+      ["Clear configuration state"]
+      (unit_p @-> returning unit_p err)
+
+  (* external reset_state: unit -> unit = "" *)
+  let reset_state = declare
+      "reset_state"
+      ["Reset configuration state"]
+      (unit_p @-> returning unit_p err)
+
+  (* external set_gateway_interface: debug_info -> name:iface -> unit = "" *)
+  let set_gateway_interface =
+    let name_p = Param.mk ~name:"name" ~description:["gateway name"] iface in
+    declare
+      "set_gateway_interface"
+      ["Set gateway interface"]
+      (debug_info_p @-> name_p @-> returning unit_p err)
+
+  (* external set_dns_interface: debug_info -> name:iface -> unit = "" *)
+  let set_dns_interface =
+    let name_p = Param.mk ~name:"name" ~description:["gateway name"] iface in
+    declare
+      "set_dns_interface"
+      ["Set dns interface"]
+      (debug_info_p @-> name_p @-> returning unit_p err)
+
+  (* module Interface = struct *)
+  module Interface = struct
+    let iface_name_p = Param.mk ~name:"name" ~description:["interface name"] iface
+
+    (* 	external get_all : debug_info -> unit -> iface list = "" *)
+    let get_all =
+      let module T = struct
+        type _iface_list_t = iface list [@@deriving rpcty]
```

Comment:
@jonludlam why do we need this kind of type juggling for polymorphic types like lists or options?

Reply:
There should be a convenience approach to handle this stuff, but I cannot find it, need Jon's help.

Reply:
I am not sure there is, that's why I wanted to double check